server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -301,21 +301,7 @@ protected void doClose() {
     }
 
     public void executeDfsPhase(ShardSearchRequest request, SearchTask task, ActionListener<SearchPhaseResult> listener) {
-        rewriteShardRequest(request, new ActionListener<ShardSearchRequest>() {
-            @Override
-            public void onResponse(ShardSearchRequest request) {
-                try {
-                    listener.onResponse(executeDfsPhase(request, task));
-                } catch (Exception e) {
-                    onFailure(e);
-                }
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                listener.onFailure(e);
-            }
-        });
+        rewriteShardRequest(request, ActionListener.map(listener, r -> executeDfsPhase(r, task)));
```

Comment:
I recently stumbled upon a subtle difference between ActionListener.wrap (called by ActionListener.map) compared to how we generally instantiate an action listener like here: there are places where we don't call `onFailure` in case `onResponse` throws (here we do , so everything is fine). This has subtle consequences depending on how the listener is used: say you use a `CountDown` and wait for a response from N nodes, if you call `countDown` on both `onFailure` and `onResponse`, you may end up counting down twice for the same node in case anything goes wrong while executing `onResponse`. Maybe an edge case, but it may be easy to miss when moving over to wrap or map whenever we wait for a response from multiple nodes.